README.md
# Data Loader #

[![Build Status](https://api.travis-ci.com/shonenkov/data-loader.svg)](https://travis-ci.com/shonenkov/data-loader)
[![Coverage Status](https://coveralls.io/repos/github/shonenkov/data-loader/badge.svg)](https://coveralls.io/github/shonenkov/data-loader)

# Installing for version 0.0.1: #
```
pip install -e git+https://github.com/shonenkov/data-loader@v0.3.0#egg=data_loader-0.3.0
```



## Starting of loading MAIL RU (see `runners/run_mail_ru.py` ):
```
from data_loader import AnswerMailRuLoader

data_folder_path = 'data'
if not os.path.exists(data_folder_path):
    os.makedirs(data_folder_path)

mail_ru = AnswerMailRuLoader(
    ids=range(0, 1000),
    save=True,
    data_folder_path=data_folder_path,
    queue_maxsize=200,
    n_processes=10,
)
mail_ru.run()
```

CHANGELOG.md
CHANGELOG
=========

0.3.0
-----
- add tests
- add mail_ru loader
- add runners

0.2.0
-----
- base class ArticleLoader
- multiprocessing
- add lenta ru
- add tests
- rename "path" to "data_folder_path"

0.1.0
-----
- fix bug with global logger
- fix display of progress
- new field "log_file_path"
- changed default TIMEOUT = 1

0.0.1
-----

- initial version


tests/conftest.py
# -*- coding: utf-8 -*-
import os
import re
from os.path import dirname, exists
from os import makedirs


import pytest

from data_loader import LentaRuLoader, BankiRuLoader


TEST_ROOT = dirname(__file__)

WORD_PATTERN = re.compile(r'\b[\w\-]+\b')


@pytest.yield_fixture(scope='module')
def lenta_ru_loader():
    yield LentaRuLoader(year=2009, save=False)


@pytest.yield_fixture(scope='module')
def banki_ru_loader():
    yield BankiRuLoader(year=2018, save=False)


@pytest.fixture(scope='session', autouse=True)
def unpack_resources():
    resources = '/tmp/test-data-loader-resources'
    if not exists(resources):
        makedirs(resources)

    os.system(f'cd {resources} && tar -xvzf {TEST_ROOT}/resources/lenta_ru/2018-07-02/lenta_ru_articles.tar.gz')
    os.system(f'cd {resources} && tar -xvzf {TEST_ROOT}/resources/banki_ru/2018-07-03/banki_ru_articles.tar.gz')


tests/test_banki_ru.py
# -*- coding: utf-8 -*-
import pytest
import asyncio
import re
from os.path import join

from bs4 import BeautifulSoup

from data_loader import BankiRuLoader
from tests.conftest import TEST_ROOT, WORD_PATTERN


@pytest.mark.asyncio
@pytest.mark.parametrize('soup_path, expected_text_path', [
    ('soup0.html', 'soup0.txt'),
    ('soup1.html', 'soup1.txt'),
    ('soup2.html', 'soup2.txt'),
    ('soup3.html', 'soup3.txt'),
    ('soup4.html', 'soup4.txt'),
])
async def test_get_article_text(soup_path, expected_text_path, banki_ru_loader):
    soup_path = join(TEST_ROOT, f'resources/banki_ru/soups/{soup_path}')
    expected_text_path = join(
        TEST_ROOT, f'resources/banki_ru/article_texts/{expected_text_path}')
    soup = BeautifulSoup(open(soup_path, 'r'), 'html.parser')
    expected_text = open(expected_text_path, 'r').read()
    text = await banki_ru_loader.get_article_text(soup)
    assert re.findall(WORD_PATTERN, text) == re.findall(WORD_PATTERN, expected_text)


@pytest.mark.asyncio
async def test_bankiru_loader():
    async def my_get_soup(url):
        return BeautifulSoup(
            open(f'/tmp/test-data-loader-resources/{url.replace("/", "-")}.html', 'r'), 'html.parser'
        )

    def _my_get_all_dates():
        return [('03', '07', '2018')]

    banki_ru = BankiRuLoader(year=2018, save=False)
    banki_ru.get_soup = my_get_soup
    banki_ru._get_all_dates = _my_get_all_dates

    await asyncio.gather(
        banki_ru.prepare(), *[banki_ru.get() for _ in range(banki_ru.n_processes)]
    )
    await asyncio.gather(
        banki_ru.load(), *[banki_ru.get() for _ in range(banki_ru.n_processes)]
    )

    assert len(banki_ru.articles) == 54
    for article in banki_ru.articles:
        assert article['text']
        assert article['date'] == '03.07.2018'
        assert article['url']
        assert article['header']


tests/test_lenta_ru.py
# -*- coding: utf-8 -*-
import pytest
import asyncio
import re
from os.path import join

from bs4 import BeautifulSoup

from data_loader import LentaRuLoader
from tests.conftest import TEST_ROOT, WORD_PATTERN


@pytest.mark.asyncio
@pytest.mark.parametrize('soup_path, expected_text_path', [
    ('soup0.html', 'soup0.txt'),
    ('soup1.html', 'soup1.txt'),
    ('soup2.html', 'soup2.txt'),
    ('soup3.html', 'soup3.txt'),
    ('soup4.html', 'soup4.txt'),
])
async def test_get_article_text(soup_path, expected_text_path, lenta_ru_loader):
    soup_path = join(TEST_ROOT, f'resources/lenta_ru/soups/{soup_path}')
    expected_text_path = join(
        TEST_ROOT, f'resources/lenta_ru/article_texts/{expected_text_path}')
    soup = BeautifulSoup(open(soup_path, 'r'), 'html.parser')
    expected_text = open(expected_text_path, 'r').read()
    text = await lenta_ru_loader.get_article_text(soup)
    assert re.findall(WORD_PATTERN, text) == re.findall(WORD_PATTERN, expected_text)


@pytest.mark.asyncio
async def test_lentaru_loader():
    async def my_get_soup(url):
        return BeautifulSoup(
            open(f'/tmp/test-data-loader-resources/{url.replace("/", "-")}.html', 'r'), 'html.parser'
        )

    def _my_get_all_dates():
        return [('02', '07', '2018')]

    lenta_ru = LentaRuLoader(year=2018, save=False)
    lenta_ru.get_soup = my_get_soup
    lenta_ru._get_all_dates = _my_get_all_dates

    await asyncio.gather(
        lenta_ru.prepare(), *[lenta_ru.get() for _ in range(lenta_ru.n_processes)]
    )
    await asyncio.gather(
        lenta_ru.load(), *[lenta_ru.get() for _ in range(lenta_ru.n_processes)]
    )

    assert len(lenta_ru.articles) == 200
    for article in lenta_ru.articles:
        assert article['text']
        assert article['date'] == '02.07.2018'
        assert article['url']
        assert article['header']


data_loader/mail_ru.py
# -*- coding: utf-8 -*-
from .base import DataExternalIDLoader


class AnswerMailRuLoader(DataExternalIDLoader):

    def __init__(self, *, ids, save, data_folder_path, queue_maxsize, n_processes):
        """
        Async multiprocessing loader of Answer Mail Ru

        :param ids: iterable object with ids
        :param save: bool, True - save data, False - don't
        :param data_folder_path: folder for load data
        :param queue_maxsize: maxsize of queue for tasks
        :param n_processes: processes count
        """
        super().__init__(
            ids=ids,
            sub_url='https://otvet.mail.ru/question/{external_id}',
            save=save,
            bs4_features='lxml',
            data_folder_path=data_folder_path,
            data_folder_deep=2,
            queue_maxsize=queue_maxsize,
            n_processes=n_processes
        )

    async def get_data(self, external_id, soup):

        title = soup.find('h1', 'q--qtext').text.strip()
        raw_comments = soup.find_all('div', 'q--qcomment medium')
        comments = []
        if raw_comments:
            comments = [q.text.strip() for q in raw_comments]

        category = soup.find('a', 'black list__title list__title')
        sub_category = soup.find('a', 'medium item item_link selected')

        if category is not None:
            category = category.text.strip()
        if sub_category is not None:
            sub_category = sub_category.text.strip()

        raw_answers = soup.find_all('div', 'a--atext atext')
        answers = []
        if raw_answers:
            answers = [a.text.strip() for a in raw_answers]

        return {'id': str(external_id), 'title': title, 'category': category, 'sub_category': sub_category,
                'comments': comments, 'answers': answers}


data_loader/banki_ru.py
# -*- coding: utf-8 -*-
from .base import ArticleLoader


class BankiRuLoader(ArticleLoader):

    NAME = 'banki'

    HOST = 'http://www.banki.ru'
    DEFAULT_URL_TEMPLATE = '/news/lenta/?d={day}&m={month}&y={year}'

    def __init__(self, year, save=True, data_folder_path=None, log_file_path=None, timeout=1, host=HOST,
                 url_template=DEFAULT_URL_TEMPLATE, queue_maxsize=1, n_processes=1):
        super().__init__(year=year, host=host, url_template=url_template, save=save, data_folder_path=data_folder_path,
                         timeout=timeout, queue_maxsize=queue_maxsize, n_processes=n_processes,
                         log_file_path=log_file_path)

    async def prepare_one_day_articles(self, soup,  year, month, day):
        """
        :return: list of dict {'url': <url>, 'header': <header>}
        """
        all_a = soup.find_all('a')
        if not all_a:
            return

        one_day_articles = []
        for a in all_a:
            if 'text-list-link' in a.attrs.get('class', []) and a['href'].startswith('/news/lenta/'):
                one_day_articles.append({
                    'url': a['href'],
                    'header': a.text,
                })

        return one_day_articles

    async def get_article_text(self, soup):
        """
        :return: str text
        """
        for a in soup.find_all('article'):
            if 'article-text' in a.attrs.get('class', []):
                return a.text


data_loader/lenta_ru.py
# -*- coding: utf-8 -*-
from .base import ArticleLoader


class LentaRuLoader(ArticleLoader):

    NAME = 'lenta'

    HOST = 'https://lenta.ru'
    DEFAULT_URL_TEMPLATE = '/news/{year}/{month}/{day}/'

    def __init__(self, year, save=True, data_folder_path=None, log_file_path=None, timeout=0,
                 host=HOST, url_template=DEFAULT_URL_TEMPLATE, queue_maxsize=200, n_processes=10):
        super().__init__(year=year, host=host, url_template=url_template, save=save, data_folder_path=data_folder_path,
                         timeout=timeout, queue_maxsize=queue_maxsize, n_processes=n_processes,
                         log_file_path=log_file_path)

    async def prepare_one_day_articles(self, soup,  year, month, day):
        """
        :return: list of dict {'url': <url>, 'header': <header>}
        """
        one_day_articles = []
        for div in soup.find_all('div', attrs='b-tabloid__topic_news'):
            a = div.a
            url = a['href']
            if f'/{year}/{month}/' in url:
                one_day_articles.append({
                    'url': url,
                    'header': a.text,
                })

        return one_day_articles

    async def get_article_text(self, soup):
        """
        :return: str text
        """
        text = ''
        for div in soup.find_all('div', attrs=['b-text']):
            if div['itemprop'] == 'articleBody':
                text = f'{text} {div.text}'

        return text


data_loader/base.py
# -*- coding: utf-8 -*-
import json
import asyncio
import itertools as it
from os import makedirs
from os.path import join, exists, basename, dirname
from glob import glob
from uuid import uuid4
from time import sleep
from datetime import datetime

import aiohttp
from bs4 import BeautifulSoup


class ArticleLoader:

    NAME = 'article'

    PREPARE = 'prepare'
    LOAD = 'load'
    KILL = 'kill'

    def __init__(self, *, year, host, url_template, save, data_folder_path, timeout, queue_maxsize, n_processes,
                 log_file_path=None):
        """
        Base class for loading articles from resources, that have one day articles structure.
        Needs to define methods "get_article_text" and "prepare_one_day_articles" with signature, that they have.

        :param int or str year: year for loading
        :param str host: resource host
        :param str url_template: template for url, where articles are being. example: '/news/{year}/{month}/{day}/'
        :param bool save: True/False - save json or not
        :param str data_folder_path: abs path of directory for saving data
        :param int timeout: timeout in seconds for waiting before requesting to server
        :param int queue_maxsize: maxsize of queue for tasks
        :param int n_processes: processes count
        :param optional log_file_path: abs path of file for saving logs;
            default path: /tmp/data-loader/<name><year>-<H>:<M>:<S>-<d>.<m>.<Y>.log
        """
        self.year = int(year)
        self.save = save
        self.data_folder_path = data_folder_path
        self.host = host
        self.timeout = timeout
        self.n_processes = n_processes
        self.url_template = url_template

        self.url_cache = set()
        self.articles = []
        self.queue = asyncio.Queue(maxsize=queue_maxsize)

        self.preparing_progress = 0
        self.preparing_end = 0

        self.loading_progress = 0
        self.loading_end = 0

        self.base_path = '/tmp/data-loader'
        if not exists(self.base_path):
            makedirs(self.base_path)

        name = f'{self.NAME}{year}-{datetime.now().strftime("%H:%M:%S-%d.%m.%Y")}'

        if data_folder_path:
            self.data_path = data_folder_path
        else:
            self.data_path = join(self.base_path, name)
            if not exists(self.data_path) and save:
                makedirs(self.data_path)

        self.log_path = log_file_path if log_file_path else join(
            self.base_path, f'{name}.log')

    async def prepare(self):
        print(f'\nPreparing to load {self.NAME} {self.year}...')

        all_dates = self._get_all_dates()
        self.preparing_end = len(all_dates)

        for day, month, year in all_dates:
            url = self.url_template.format(year=year, month=month, day=day)
            if url in self.url_cache:
                self.preparing_progress += 1
                continue

            await self.queue.put((self.PREPARE, (url, year, month, day)))

        for _ in range(self.n_processes):
            await self.queue.put((self.KILL, None))

    async def load(self):
        print('\nLoading...')
        self.loading_end = len(self.articles)

        for i, article in enumerate(self.articles):
            url = article['url']
            if url in self.url_cache:
                self.loading_progress += 1
                continue

            await self.queue.put((self.LOAD, (url, i)))

        for _ in range(self.n_processes):
            await self.queue.put((self.KILL, None))

    async def get(self):
        while True:
            command, data = await self.queue.get()
            if command == self.PREPARE:
                await self.get_prepare(*data)
            elif command == self.LOAD:
                await self.get_load(*data)
            else:
                break

    async def prepare_one_day_articles(self, soup,  year, month, day):
        """
        :return: list of dicts {'url': <url>, 'header': <header>}
        """
        raise NotImplementedError

    async def get_prepare(self, url, year, month, day):

        self.preparing_progress += 1

        soup = await self.get_soup(url)
        if not soup:
            await self._log('PREPARING_ERROR', f'No soup for url "{url}"')
            return

        self.url_cache.add(url)
        one_day_articles = await self.prepare_one_day_articles(soup, year, month, day)
        if not one_day_articles:
            await self._log('PREPARING_WARNING', f'Articles was not found from url "{url}"')
            return

        for one_day_article in one_day_articles:
            self.articles.append({
                'url': one_day_article['url'],
                'header': one_day_article['header'],
                'date': f'{day}.{month}.{year}',
                'text': ''
            })

        if self.preparing_progress < self.preparing_end:
            print(f'\r{self.preparing_progress} from {self.preparing_end} days', end='')
        elif self.preparing_progress == self.preparing_end:
            self.preparing_progress += 1
            print(f'\nPreparing {self.NAME} {self.year} finished!\n')

    async def get_article_text(self, soup):
        """
        :return: str text
        """
        raise NotImplementedError

    async def get_load(self, url, i):

        self.loading_progress += 1

        soup = await self.get_soup(url)
        if not soup:
            await self._log('WARNING', f'No soup for url "{url}"')
            return

        text = await self.get_article_text(soup)
        if not text:
            await self._log('WARNING', f'No text for url "{url}"')
            return

        self.articles[i]['text'] = text

        if self.save:
            name = f'{uuid4()}.json'
            self._save_json(self.articles[i], name)
            await self._log('INFO', f'Article with url "{url}" was saved as "{name}".')

        if self.loading_progress < self.loading_end:
            print(f'\r{self.loading_progress} from {self.loading_end} articles', end='')
        elif self.loading_progress == self.loading_end:
            self.loading_progress += 1
            print(f'\nLoading {self.NAME} {self.year} finished!\n')

    async def get_soup(self, url):
        sleep(self.timeout)
        try:
            async with aiohttp.ClientSession() as session:
                raw_response = await session.get(url=f'{self.host}{url}')
                if raw_response.status != 200:
                    print('Bad status! Use correct timeout.')
                    await self._log('ERROR', f'Bad status from server! "{url}"')
                    return
                content = await raw_response.content.read()

            self.url_cache.add(url)
            return BeautifulSoup(content, 'html.parser')
        except Exception as e:
            await self._log('ERROR', f'{e, type(e)}')

    async def _log(self, status, msg):
        with open(self.log_path, 'a+') as logger:
            logger.write(f'{status}:{msg}\n')

    def _get_all_dates(self):
        all_dates = []
        for day, month, year in it.product(range(1, 32), range(1, 13), range(self.year, self.year + 1)):
            try:
                if datetime(year=year, month=month, day=day) > datetime.now():
                    raise ValueError
            except ValueError:
                continue
            day = str(day).zfill(2)
            month = str(month).zfill(2)
            year = str(year)
            all_dates.append([day, month, year])
        return all_dates

    def _save_json(self, article, name):
        file = open(join(self.data_path, name), 'w')
        json.dump(article, file, ensure_ascii=False)
        file.close()

    def run(self):
        """ Main method, after finished you can find data in self.articles or in folder with path <path> """
        loop = asyncio.get_event_loop()
        loop.run_until_complete(asyncio.gather(
            self.prepare(), *[self.get() for _ in range(self.n_processes)]))
        loop.run_until_complete(asyncio.gather(
            self.load(), *[self.get() for _ in range(self.n_processes)]))


class DataExternalIDLoader:

    NAME = 'data_external_id'

    LOAD = 'load'
    KILL = 'kill'

    def __init__(self, *, ids, sub_url, save, bs4_features, data_folder_path, data_folder_deep, queue_maxsize,
                 n_processes, log_file_path=None):
        """
        Base class for loading data from external_id, that have one page structure.
        :param ids: iterable object with ids
        :param sub_url: template url with external_id, example: 'https://otvet.mail.ru/question/{external_id}'
        :param save: bool, True - save data, False - don't
        :param bs4_features: 'lxml'/'html'
        :param data_folder_path: folder for load data
        :param data_folder_deep: 1000**n, n - count of separation data
        :param queue_maxsize: maxsize of queue for tasks
        :param n_processes: processes count
        :param log_file_path: default as {NAME}.log
        """

        self.data_folder_path = data_folder_path
        self.data_folder_deep = data_folder_deep
        self.bs4_features = bs4_features
        self.n_processes = n_processes
        self.ids = ids
        self.loaded_external_ids = {basename(path)[:-5] for path in glob(
            join(data_folder_path, *['*' for _ in range(data_folder_deep)], '*.json')
        )}
        self.save = save
        self.loading_progress = 0
        self.queue = asyncio.Queue(maxsize=queue_maxsize)
        self.sub_url = sub_url
        self.log_path = log_file_path if log_file_path else f'{self.NAME}.log'

    async def load(self):
        print('\nLoading...')

        for external_id in self.ids:
            print(self.loading_progress, end='\r')
            if external_id in self.loaded_external_ids:
                self.loading_progress += 1
                continue
            await self.queue.put((self.LOAD, (external_id,)))

        for _ in range(self.n_processes):
            await self.queue.put((self.KILL, None))

    async def get(self):
        while True:
            command, data = await self.queue.get()
            if command == self.LOAD:
                await self.get_load(*data)
            else:
                break

    async def get_data(self, external_id, soup):
        """ Method for parse data from soup. Return JSON  """
        raise NotImplementedError

    async def get_load(self, external_id):

        self.loading_progress += 1

        soup = await self.get_soup(external_id)
        if not soup:
            await self._log('WARNING', f'No soup for external_id "{external_id}"')
            return

        data = await self.get_data(external_id, soup)
        if not data:
            await self._log('WARNING', f'No text for external_id "{external_id}"')
            return

        if self.save:
            name = [f'{external_id}.json']
            for _ in range(self.data_folder_deep):
                name.append(str(external_id // 1000 % 1000))
                external_id = external_id // 1000

            path = join(*name[::-1])
            sub_folder_path = dirname(join(self.data_folder_path, path))
            if not exists(sub_folder_path):
                makedirs(sub_folder_path)

            self._save_json(data, path)
            await self._log('INFO', f'Data saved as "{path}".')

    async def get_soup(self, external_id):
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(url=self.sub_url.format(external_id=external_id), verify_ssl=False) \
                        as raw_response:
                    if raw_response.status == 404:
                        await session.close()
                        return
                    elif raw_response.status != 200:
                        print('Bad status! Use correct timeout.')
                        await self._log('ERROR', f'Bad status from server! "{external_id}"')
                        await session.close()
                        return
                    content = await raw_response.content.read()

            self.loaded_external_ids.add(external_id)
            return BeautifulSoup(content, self.bs4_features)
        except Exception as e:
            await self._log('ERROR', f'{e, type(e)}')

    async def _log(self, status, msg):
        with open(self.log_path, 'a+') as logger:
            logger.write(f'{status}:{msg}\n')

    def _save_json(self, article, name):
        file = open(join(self.data_folder_path, name), 'w')
        json.dump(article, file, ensure_ascii=False)
        file.close()

    def run(self):
        """ Main method """
        loop = asyncio.get_event_loop()
        loop.run_until_complete(asyncio.gather(
            self.load(), *[self.get() for _ in range(self.n_processes)]))
